### PR TITLE
Z-score signal amplification 

### PR DESCRIPTION
**Pull request by alchemystack**
Replace old mode-base token sampling in LLM inference with a z-score aggregate process designed to allow better consciousness influence over token selection. The algorithm converts 20,480 raw INT8 samples from a quantum random number generator (QRNG) into a single uniform float in [0, 1), which is then used for standard CDF-based token sampling against the softmax distribution.

The statistical amplification works by aggregating thousands of individually weak quantum measurements into a single mean value, where even a sub-0.2% per-sample bias from consciousness produces a detectable shift in token selection.

This is now default and changes the color coding to blue = negative z-score = high probability tokens / red = positive z-score = lower probability tokens. (this is WIP)

### Input

- `logits`: raw logit vector from the LLM, shape `[vocab_size]`
- `temperature`: sampling temperature (float > 0)
- `qrng_samples`: 20,480 unsigned 8-bit integers from hardware QRNG
- `top_k` (optional): if set, restrict to top-k tokens before sampling
- `top_p` (optional): if set, restrict to nucleus of tokens with cumulative probability ≤ top_p

### Step 1 – Prepare the token distribution

1. Apply temperature scaling: `scaled_logits = logits / temperature`
2. If `top_k` is set, zero out all logits outside the top-k highest values (set to -∞)
3. Compute softmax probabilities: `probs = softmax(scaled_logits)`
4. If `top_p` is set, sort tokens by descending probability, find the smallest set whose cumulative probability exceeds `top_p`, zero out the rest, and renormalize

### Step 2 – Build the probability-ordered CDF

This ordering is essential. It gives the consciousness influence lever a coherent meaning: higher quantum signal → less probable (more surprising) token selection.

1. Create an index array of all tokens with nonzero probability
2. Sort these indices by their probability in **descending** order (highest probability first)
3. Build the cumulative distribution function over this sorted order:
   - `sorted_probs = [probs[idx] for idx in sorted_indices]`
   - `cdf[0] = sorted_probs[0]`
   - `cdf[i] = cdf[i-1] + sorted_probs[i]` for i > 0
4. The final CDF value should equal 1.0 (renormalize if needed due to floating point)

### Step 3 – Convert QRNG samples to uniform float

Constants (derived from the uniform distribution on 0–255):

- Population mean: `μ = 127.5`
- Standard error of the mean for n=20480 samples: `σ_m = 73.6116 / √20480 ≈ 0.51433`

Where `73.6116` is the standard deviation of a single discrete uniform variable on [0, 255], computed as `√((256² - 1) / 12)`.

Procedure:

1. Compute the sample mean:
   ```
   M = sum(qrng_samples) / 20480
   ```

2. Compute the z-score:
   ```
   z = (M - 127.5) / 0.51433
   ```

3. Convert to uniform via the standard normal CDF (Φ):
   ```
   u = Φ(z) = 0.5 × (1 + erf(z / √2))
   ```

4. Clamp to avoid degenerate edge values:
   ```
   u = clamp(u, 1e-10, 1 - 1e-10)
   ```

The value `u` is now a float in (0, 1). Under the null hypothesis (no consciousness influence), `u` is uniformly distributed. This is guaranteed by the probability integral transform.

### Step 4 – Select the token

1. Find the smallest index `k` in the sorted CDF such that `cdf[k] ≥ u`
2. The selected token ID is `sorted_indices[k]`

### Interpretation of the consciousness lever

Because the CDF is ordered from most probable to least probable:

- `u` close to 0 → selects high-probability tokens (conservative, expected completions)
- `u` close to 1 → selects low-probability tokens (surprising, creative completions)
- Consciousness shifts the mean of INT8 samples upward → higher `u` → less obvious token
- Consciousness shifts the mean downward → lower `u` → more conventional token

A shift of the sample mean by just 0.43 (from 127.5 to 127.93) produces u ≈ 0.80, which is enough to skip past the most likely tokens into the tail of the distribution.

## Pseudorandom compatibility

For controlled experiments comparing quantum vs. pseudorandom sampling, replace step 3 with:

1. Generate a single FP32 uniform random value `u` from a PRNG in [0, 1)
2. Proceed to step 4 as normal

The rest of the pipeline is identical. The z-score path and the direct PRNG path both produce a uniform float consumed the same way, making A/B comparison straightforward.

## Key constants

| Parameter | Value | Derivation |
|---|---|---|
| QRNG sample count (n) | 20,480 | Chosen for QRNG batch efficiency |
| INT8 range | [0, 255] | Unsigned 8-bit |
| Population mean (μ) | 127.5 | (0 + 255) / 2 |
| Population std (σ) | 73.6116 | √((256² - 1) / 12) |
| Std error of mean (σ_m) | 0.51433 | σ / √n |
| Min detectable bias | ~0.2% per sample | Shifts mean by ~0.5σ_m |

### CODE REVIEW

**Comment by alchemystack** [comment]
Sorry, wrong repo, meant to merge with my own fork 😅
